tests/cases/fourslash/completionsOptionalReplacementSpan1.ts

```diff
@@ -0,0 +1,13 @@
+/// <reference path="fourslash.ts" />
+
+// @lib: dom
+
+//// console.[|cl/*0*/ockwork|];
+//// type T = Array["[|toS/*1*/paghetti|]"];
```

Comment:
I am not sure why this has to be optionalReplacementSpan and not just replacementSpan if editor supports it (there is already preference for that no)

Reply:
My understanding is there is not a preference that allows users to opt out of using `replacementSpan`—in VS and VS Code, the presence of `replacementSpan` means the editor will always use it, which is good because it means users can’t end up in a situation where they trigger completions on `style.|` and end up with a syntax error like `style.["background-color"]`. This new span will be configurable by a preference.

That said, it does seem like using the new `optionalReplacementSpan` is almost _always_ better than the current behavior, but I guess @mjbvz and co. have thought through the reasons someone might want to have both options more than I have.

Reply:
But i dont see this being configured by preference? Even if this is configured by preference, why cant it be returned as part of replacementSpan since replacement span if present is always meant to override this one?

Reply:
I believe it’s supposed to be an editor preference, which could potentially be configured across multiple languages. If TS Server adopted this preference and VS Code sent it to us, that would be an option. That’s not something that was discussed in the issue; I personally don’t have strong feelings one way or another. I’ll let @mjbvz weigh in.

Reply:
VS code lets users configure how suggestions are inserted using the `editor.suggest.insertMode` setting. The options are `insert` which always inserts the current suggestion without replacing the suffix, and `replace` which would try replacing the suffix word.

However users can always toggle to the opposite mode of the current one by holding down `shift` while accepting a completion. This means that TS always needs to return both spans to us in some way. I think we need a clear distinction between: 

-Here's a span that should always be replaced regardless of insert mode (the current `replacementSpan` property)
- Here's a replacement span that you should use if the user accepts a completion in `replace` mode (which would be the new `optionalRepalcementSpan`)

I think the current proposal covers our needs but let me know if that doesn't sound correct based on my description

Reply:
Got it. Now understand why you always want this property to be populated and separate. Thanks.

Reply:
Thanks for the clarification 👍